File: 2021/JavaScript/AOC-2021-02.js
let aoc = require('./AOC-2021.js');

let solver = new aoc.Solver('02', '2021');

solver.parseInput = function (input) {
  return input
    .trim()
    .split("\n")
    .map((i) => { 
      var instr = i.split(' '); return { 
        direction: instr[0], 
        distance: parseInt(instr[1]) 
      } 
    });
};

solver.solvePart1 = function (data) {
  var pos = { depth: 0, dist: 0 };
  for(var d = 0; d < data.length; d++) {
    switch(data[d].direction) {
      case "forward":
        pos.dist += data[d].distance;
        break;
      case "up":
        pos.depth -= data[d].distance;
        break;
      case "down":
        pos.depth += data[d].distance;
        break;
    }
  }
  return pos.depth * pos.dist;
};

solver.solvePart2 = function (data) {
  var pos = { depth: 0, dist: 0, aim: 0 };
  for(var d = 0; d < data.length; d++) {
    switch(data[d].direction) {
      case "forward":
        pos.dist += data[d].distance;
        pos.depth += pos.aim * data[d].distance;
        break;
      case "up":
        pos.aim -= data[d].distance;
        break;
      case "down":
        pos.aim += data[d].distance;
        break;
    }
  }
  return pos.depth * pos.dist;
};

aoc.run(solver);

File: 2021/Rust/src/AOC-2021-02.rs
use std::{
    fs::File,
    io::{self, BufRead, BufReader},
    path::Path,
};

const YEAR: i32 = 2021;
const DAY: i32 = 2;

struct Instruction {
    direction: String,
    distance: i32
}
    
fn lines_from_file(filename: impl AsRef<Path>) -> io::Result<Vec<String>> {
    BufReader::new(File::open(filename)?).lines().collect()
}

fn parse_input(s: Vec<String>) -> Vec<Instruction> {
    return s.iter().map(|x| create_instruction(x)).collect() 
}

fn create_instruction(s: &String) -> Instruction {
    let s: Vec<&str> = s.split(' ').collect();
    return Instruction { 
        direction: s[0].to_string(), 
        distance: s[1].parse::<i32>().unwrap()
    };
}

fn find_answer_1(data: &Vec<Instruction>) -> i32 {
    let mut depth = 0;
    let mut distance = 0;
    for(_i, item) in data.iter().enumerate() {
        match item.direction.as_str() {
            "forward" => { distance += &item.distance; }
            "up" => { depth -= &item.distance; }
            "down" => { depth += &item.distance; }
            _ => println!("instruction mismatch")
        }
    }
    return depth * distance;
}

fn find_answer_2(data: &Vec<Instruction>) -> i32 {
    let mut depth = 0;
    let mut distance = 0;
    let mut aim = 0;
    for(_i, item) in data.iter().enumerate() {
        match item.direction.as_str() {
            "forward" => { 
                distance += &item.distance;
                depth += &item.distance * aim;
            }
            "up" => { aim -= &item.distance; }
            "down" => { aim += &item.distance; }
            _ => println!("instruction mismatch")
        }
    }
    return depth * distance;
}

fn main() {
    let filename = format!("..\\..\\scratch\\{}-{:0>2}-input.txt", YEAR, DAY);

    println!("AOC {}/{:0>2}", YEAR, DAY);

    let data = lines_from_file(&filename)
        .expect(format!("Unable  to read {}", &filename).as_str());

    let input = parse_input(data);

    println!("Part 1 Answer: {}", find_answer_1(&input));

    println!("Part 2 Answer: {}", find_answer_2(&input));
}